### PR TITLE
Replace build image bash script  in CI with build image in python

### LINKED ISSUE #20741: CI: Reuse Breeze CI image building for CI
We already have a basic Build image function in `Breeze` - we should be able to use it to build images in CI.

There are actions that build images in `./build-imaages.yml` and they should simply (similarely to free-space) use Python commands develped in ./dev/breeze.

### PR DESCRIPTION
**Pull request by edithturn**
The script to build image is already created in python: build_image.py.
This PR replaces ./dev/breeze/src/airflow_breeze/ci/ci_build_ci_image_on_ci.sh with build_image.py
Closes #20741

### CODE REVIEW

**Comment by edithturn** [comment]
@potiuk these are what I did and the status of this PR.

* There were many things that were synchronized with the PR of @Bowrna 👏🏼 
* In Breeze.py I am taking two-variable environments: **GITHUB_REPOSITORY** and **PYTHON_MAJOR_MINOR_VERSION**.
* Replacing ./scripts/ci/images/ci_build_ci_image_on_ci.sh for Breeze2 build-ci-image --verbose (verbose just for revision)
* With these changes, BUILD IMAGE works, but the problem is with Build PROD image 3.7, **which always takes more time and fails. How I can test with this, look like time out.**  You can see it here:  https://github.com/edithturn/airflow/runs/5430496264?check_suite_focus=true

Note: There was a random error when the image was created and then pushed, it was failing the compilation for python 3.7, 3.8 y 3.9 (randomly), the Job to Push was asking permissions to write on the package after this was built, which means the packages on my repository weren't able to overwrite. I changed the configuration of "packages" giving "write access" to each package version, it is how start to work on my own repository. You can see it here: https://github.com/potiuk/airflow/actions/runs/1919547974


I checked that the ci.yaml also has the ci_build_ci_image_on_ci.sh, should we also replace that too?
https://github.com/apache/airflow/blob/f5f221605632ebb938907bb6586537c81baa066e/.github/workflows/ci.yml#L1403-L1404

**Comment by potiuk** [comment]
This is really cool and almost works :) 

Few things: 

1) rebase please :) 
2) please take a look at all the options and corresponding variables and update all the options to use the variables from the original `breeze`: you will see all the --flags and corresponding variables here: https://github.com/apache/airflow/blob/8c63a8d3cdc6c6d3911a8e7bde63f4fae1dde67c/breeze#L866 
3) I think what you miss currently, is to make package 3.7 publicly available - you need to make sure that the package has the right access:

* It has to be public 
* It has to have "airflow" repository with "write" access for Github actions

Some screenshots from my https://github.com/users/potiuk/packages/container/airflow%2Fmain%2Fci%2Fpython3.7/settings

Actions write access:

Screenshot 2022-03-05 11 26 52

The package has to be public. Originally it is private but you can change visibility and once you do it, it should look like this: 

Screenshot 2022-03-05 11 30 04

**Comment by potiuk** [comment]
BTW. if you want @edithturn  you can add me as admin to your fork, then I will be able to view and edit access and help:

image

**Comment by edithturn** [comment]
@potiuk I updated the options with the environment variables declared on breeze, It looks like it is working fine, you can check it here: https://github.com/edithturn/airflow/actions/runs/1949750121 , 

I rebased. I am working in my main to make it possible to see build_image. Please could you check it here, first: https://github.com/edithturn/airflow/actions/runs/1949750121 , then I will update this branch.

According to what I understood, the parameters that are being sent to build-ci-image are the ones necessary for the image to be built. There are others that are optional but we are not considering them yet.

At the start, I was adding all the breeze variables, but realized they wouldn't be used yet. I just added envvar in the options of the parameters that are passed in the function.

* Just in case: You are admins of my repository

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
We are getting really close to get `build-ci-image` functinality to be removed from the old Breeze @bowrna @edithturn . Once this once this one is merged, I am going to replace the old breeze's buld image with this one. :)

**Comment by Bowrna** [comment]
> We are getting really close to get `build-ci-image` functinality to be removed from the old Breeze @Bowrna @edithturn . Once this once this one is merged, I am going to replace the old breeze's buld image with this one. :)

Can you share about how you are planning to replace the old breeze's build image with this one. @potiuk

**Comment by edithturn** [comment]
@potiuk I saw **build_cache: Optional[str],** is not associated to any variable environment unless this is: build_cache_local what we mean by build_cache, it would make sense to change it to --build-cache-local which is a boolean value.

**Comment by edithturn** [comment]
@potiuk could you please execute with your permissions to see if build_image is working fine with the update?

**Comment by potiuk** [comment]
> @potiuk I saw **build_cache: Optional[str],** is not associated to any variable environment unless this is: build_cache_local what we mean by build_cache, it would make sense to change it to --build-cache-local which is a boolean value.

This does not have to be set. We use `--build-cache-local` only for some very specific cases where we want to iterate over the build locally and we do not want to use remote cache. In CI we always use the default (pulled). So we can leave that one

**Comment by potiuk** [comment]
> @potiuk could you please execute with your permissions to see if build_image is working fine with the update?

I can't do it :) . It only works if it is pushed to `main` so we won't be able to see it working in `airflow` repo until we merge :scream: .

File: breeze
Comment (potiuk) [inline breeze:839]: Nice!

**Review by potiuk — COMMENTED** [review]
This looks cool. I will push it to my main to double-check and I think it should be ready to merge

**Comment by edithturn** [comment]
😱 
Also, I was thinking of having a development environment, where we can test this kind of thing before passing it to the main.
Like this branches:

- main
- dev
- features

**Comment by potiuk** [comment]
> Can you share about how you are planning to replace the old breeze's build image with this one. @potiuk

I plan to simply run `./Breeze2 build-ci-image` command directly in the old `breeze` script. Just make people switch to it without even noticing :scream: :scream: :scream: .... eventually when we replace all functionality we will just replace the old `./breeze` script with `./Breeze2` :).

That will not yet work with autocomplete from the new `Breeze2` (still the ./breeze one will be used) but 
* we still have to fix the scenario of installing of the autocomplete with the `pipx` and virtualenv 
* we will have a chance to check if things are working for everyone else one-functionality-by-one-functionality without "big-bang"

**Comment by potiuk** [comment]
> Also, I was thinking of having a development environment, where we can test this kind of thing before passing it to the main. Like this branches:
> 
> * main
> * dev
> * features

That can be done easily with forks as we do now. We have another idea (but it is a bit more tricky to implement. The only reason we have this separation of code executed in build-image and ci is security (we do not want "casual" users to run builds using our self-hosted runners - they could do nasty stuff (like mining bitcoin using our machines). But we can make it so when PR comes from the maintainer (trusted user), we allow to use the code coming from PR in build.yml. This shoudl be possible but it is a bit tricky to implement,

**Comment by edithturn** [comment]
I will try to do it : `git push edithturn reuse-breeze-CI-image-building-for-CI:main` 
I forgot to remove the --verbose from  `Breeze2 build-ci-image --verbose` 😶‍🌫️

**Comment by potiuk** [comment]
Actually - that is a good point. We WANT --verbose in CI. But it should also be controlled by `VERBOSE` env - so please remove `--verbose` as flag, but add it as envvar - so that it will still be `verbose`.

**Comment by potiuk** [comment]
:tada: :tada: :scream: :scream: :scream: :scream:

**Comment by edithturn** [comment]
Thanks, @potiuk for all your support. 🤸🏼‍♀️

**Pull request merged**
